File: handy-ts/quiz/q2_compress_string_easy.ts
/**
 * Q2) Compress String (Run-Length Encoding)
 *
 * AI-BEST:
 * Build chunks in array and join once.
 * Time: O(n), Space: O(n)
 *
 * AI-EASY:
 * Build result by direct string concatenation.
 * Time: can degrade to O(n^2), Space: O(n)
 */

function compressStringBest(s: string): string {
  if (!s) return "";

  const chunks: string[] = [];
  let count = 1;

  for (let i = 1; i <= s.length; i++) {
    if (i < s.length && s[i] === s[i - 1]) {
      count++;
    } else {
      chunks.push(`${s[i - 1]}${count}`);
      count = 1;
    }
  }

  const compressed = chunks.join("");
  return compressed.length < s.length ? compressed : s;
}

function compressStringEasy(s: string): string {
  if (!s) return "";

  let result = "";
  let count = 1;

  for (let i = 1; i <= s.length; i++) {
    if (i < s.length && s[i] === s[i - 1]) {
      count++;
    } else {
      result += `${s[i - 1]}${count}`;
      count = 1;
    }
  }

  return result.length < s.length ? result : s;
}

function runDemo(): void {
  console.log("Q2: Compress String");
  const samples = ["aabcccccaaa", "abc", "", "aaAA"];
  for (const s of samples) {
    console.log(`Input: "${s}"`);
    console.log("  BEST:", compressStringBest(s));
    console.log("  EASY:", compressStringEasy(s));
  }
}

runDemo();

export {};

File: handy-python/quiz/q2_compress_string_easy.py
"""
Q2) Compress String (Run-Length Encoding)

AI-BEST:
- Use list builder then "".join once.
- Time: O(n), Space: O(n)

AI-EASY:
- Build output via direct string concatenation.
- Time: can degrade to O(n^2), Space: O(n)
"""


def compress_string_best(s: str) -> str:
    if not s:
        return ""

    parts: list[str] = []
    count = 1

    for i in range(1, len(s) + 1):
        if i < len(s) and s[i] == s[i - 1]:
            count += 1
        else:
            parts.append(f"{s[i - 1]}{count}")
            count = 1

    compressed = "".join(parts)
    return compressed if len(compressed) < len(s) else s


def compress_string_easy(s: str) -> str:
    if not s:
        return ""

    result = ""
    count = 1

    for i in range(1, len(s) + 1):
        if i < len(s) and s[i] == s[i - 1]:
            count += 1
        else:
            result += f"{s[i - 1]}{count}"
            count = 1

    return result if len(result) < len(s) else s


def run_demo() -> None:
    print("Q2: Compress String")
    samples = ["aabcccccaaa", "abc", "", "aaAA"]
    for s in samples:
        print(f'Input: "{s}"')
        print("  BEST:", compress_string_best(s))
        print("  EASY:", compress_string_easy(s))


if __name__ == "__main__":
    run_demo()
